torch/_functorch/aot_autograd.py

```diff
@@ -1340,10 +1343,10 @@ def _aot_export_function(
         decompositions=decompositions,
         num_params_buffers=num_params_buffers,
         aot_id=next(AOT_COUNTER),
-        # For now there's no use case involving keeping input mutations in the graph
-        # (which we can only do in the inference case anyway).
-        # We can add this later if we need to.
-        keep_inference_input_mutations=False,
+        # Regarding aten operations with `out` parameter, we need to enable
+        # keep_inference_input_mutations to guarantee the correctness as
+        # the `out` parameter needs to be mutated.
+        keep_inference_input_mutations=keep_inference_input_mutations,
```

Comment:
I am not sure if there is any implication to functionalization, cc @bdhirsh

Reply:
hmm... I am a bit curious, since I think you can get input mutations working without this change for export.

The normal flow with export (as I understand it) is that you call `export()` to get out a purely functional graph, and you can then call an `unlift()`-style function (or `ExportedProgram.module()`) to re-introduce the mutations into the graph.

So is there any reason you need to make the change here, instead of using those existing API's?

Reply:
@bdhirsh , thanks for the information. I just missed the `unlift`-style function and will check it. Regarding this change, it is concluded from the comments - https://github.com/pytorch/pytorch/pull/124926/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416L1343
 > For now there's no use case involving keeping input mutations in the graph (which we can only do in the inference case anyway). We can add this later if we need to.

But `unlift` is the best way to support it. Thanks, @bdhirsh

Reply:
yeah - that is entirely fair, this comment is just out of date :)

Reply:
@bdhirsh , I checked the `unlift` logic. It requires the input must be in the `inputs_to_parameters` or `inputs_to_buffers`. Regarding AOTI Eager, it only compiles aten operations w/o any buffers and parameters. So, it may not be able to utilize `unlift` to support the `out` parameter of aten operations directly, like `aten::clamp.out(Tensor self, Scalar? min=None, Scalar? max=None, *, Tensor(a!) out) -> Tensor(a!)`. Is my understanding correct?

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L167-L180
```python
    for node in placeholder_nodes:
        node_name = node.name
        if node_name in graph_signature.inputs_to_parameters:
            parameter_name = graph_signature.inputs_to_parameters[node_name]
            lifted_inputs.append(parameter_name)
        elif node_name in graph_signature.inputs_to_buffers:
            buffer_name = graph_signature.inputs_to_buffers[node_name]
            lifted_inputs.append(buffer_name)
            gm.meta[
                get_cloned_parameter_buffer_name(buffer_name)
            ] = clone_preserve_strides(state_dict[buffer_name])
        else:
            assert node_name in graph_signature.user_inputs
            lifted_inputs.append(None)
```

Reply:
Hmm @EikanWang I talked to Bin, and:

(1) I would prefer not to add the extra config / differences in AOTAutograd directly: right now, export expects a fully functional graph, and the general contract is that we return a functional graph here, and you can add the input mutations back in a post-processing step later..

(2) I think it would be reasonable to update the `unlift()` logic you linked to also add the input mutations back, when those input mutations happen to graph inputs (and not buffers/params).

Reply:
Thanks for your clarification. I will remove the `keep_input_mutation` related logic.